### PR TITLE
Distance floor

### PR DESCRIPTION
**Pull request by Nmargolis**
# Distance floor PR

## Context

In calculating scoredist, the distratio had a [hard-coded minimum value of 0.005](https://github.com/mapbox/carmen-cache/blob/master/src/cpp_util.cpp#L158-L161). The rationale behind this was that if as distratio approaches zero, the scordist value gets exponentially higher. The cap at 0.05 was hard-coded based on a proximity radius of 200, assuming that 0.05 was the minimum reasonable value (more explanation below).

This value of 0.005 is based on the minimum reasonable value for a proximity radius closer to 200. The problem is that when the proximity radius passed in to coalesce is higher, say 400, there is meaningful differentiation in the values that are less than 0.005.

Alternatively, any distance less than 1 isn't meaningful, since something on the same tile as the proximity point will be 0, and something one tile away will have a distance of >=1.

This PR sets a minimum distance, instead of a minimum distRatio.

## Summary of changes
- [x] Give distance a minimum value to avoid distRatios that approach zero, and exponentially high scoredists

## Next steps
- [x] More comprehensive analysis of different zoom levels, proximity radius values, and scores
- [x] Potentially evaluate whether something like 0.9 would make more sense than 0.5

## More in-depth explanation

[**zoomTileRadius**](https://github.com/mapbox/carmen-cache/blob/master/src/cpp_util.cpp#L104-L110) converts miles into a tile unit value. For zoom 14, it would be `0.8`:

```
    // 32 tiles is about 40 miles at z14, use this as our mile <=> tile conversion.
    return (32.0 / 40.0) / _pow(1.5, 14 - static_cast<int>(zoom)); // 0.8 / 1.5^0 = 0.8
```
[**proximityRadius**](https://github.com/mapbox/carmen-cache/blob/master/src/cpp_util.cpp#L113-L129) for zoom 14, and radius 400 would be `320` (as opposed to `160` for radius 200):

```
    if (zoom >= 6 && zoom <= 14) {
        return radius * zoomRadius[14 - zoom]; // 400 * 0.8 = 320
    }
```

[**distRatio**](https://github.com/mapbox/carmen-cache/blob/master/src/cpp_util.cpp#L156-L165) for zoom 14, distance 0, score 0, radius 400 would previously have been capped at 0.005:

**Previously:**

```
    double distRatio = distance / proximityRadius(zoom, radius); 
    // For a distance of 0 and radius of 400: 0 / 320 = 0
    // For a distance of 1 and radius of 400: 1/320 = 0.003125
    
    // Too close to 0 the values get intense. Cap it.
    if (distRatio < 0.005) {
        distRatio = 0.005;
    }
    // Both would return 0.005, resulting in no differentiation
```
If the feature was 1 tile away from the proximity point, with a radius of 200 (so proximityRadius of 160), the distRatio woud have been `1/160 = 0.00625`. That’s why a minimum of `0.005` made sense. If the proximity radius is 400 though, the distRatio for 1/320 would be  `0.003125`, which is under the 0.005 threshold. This means that there was no distinction in scoredist between a distance of 0 and a distance of 1, so features on the same tile as the proximity point were not getting sorted higher than features 1 tile away.

**Now:**

```
    // Too close to 0 the scoredist values get intense. Give distance a floor.
    if (distance < 1) {
        // Something greater than 0 but less than 1, to avoid dividing by 0
        distance = 0.5;
    }
    double distRatio = distance / proximityRadius(zoom, radius); 
    
    // For a distance of 0, corrected to 0.05: 0.5 / 320 = 0.0015625
    // For a distance of 1: 1/320 = 0.003125
    // There is meaningful differentiation in these values
   ``` 

[**scoreDist**](https://github.com/mapbox/carmen-cache/blob/master/src/cpp_util.cpp#L167)**:** 
```
    static const double E_POW[8] = {
        1,
        2.718281828459045,
        7.38905609893065,
        20.085536923187668,
        54.598150033144236,
        148.4131591025766,
        403.4287934927351,
        1096.6331584284585};
    
    // distratio calculations ...
    
    return ((6 * E_POW[score] / E_POW[7]) + 1) / distRatio;
    // For score 0, distRatio 0.0015625: ((6 * 1) / 1096.63) + 1) / 0.0015625 = ~643.50163
    // For score 0, distRatio 0.003125: ((6 * 1) / 1096.63) + 1) / 0.003125 = ~321.75081
```

## Open question:

What are the limits of this for different zoom levels, scores and proximity radius values? Is it still necessary to have another cap to avoid astronomically high scoredists?

### CODE REVIEW

File: src/cpp_util.cpp [hunk 1]
```diff
@@ -153,14 +153,16 @@ double scoredist(unsigned zoom, double distance, unsigned short score, double ra
...
+
     double distRatio = distance / proximityRadius(zoom, radius);
 
-    // Too close to 0 the values get intense. Cap it.
-    if (distRatio < 0.005) {
-        distRatio = 0.005;
-    }
     // Beyond the proximity radius just let scoredist be driven by score.
-    else if (distRatio > 1.0) {
+    if (distRatio > 1.0) {
```
Comment (miccolis) [inline src/cpp_util.cpp]: @Nmargolis why did you remove the else here? These conditions are still exclusive right?

File: test/coalesce.test.js [hunk 2]
```diff
@@ -302,14 +302,14 @@ test('coalesce args', (t) => {
                 cache: cache,
                 mask: 1 << 0,
                 idx: 0,
-                zoom: 2,
+                zoom: 14,
```
Comment (miccolis) [inline test/coalesce.test.js]: I agree that running these tests at zoom 2 is totally unrealistic so :+1: on that change. Not sure I follow why some are 14 and some are 6, should we change something about them so that they can all be 6?

File: test/coalesce.test.js [hunk 3]
```diff
@@ -291,7 +291,7 @@ test('coalesce args', (t) => {
             x: 1,
             y: 1,
             relev: 1,
-            score: 7
+            score: 3
```
Comment (miccolis) [inline test/coalesce.test.js]: I'm a little worried that we're losing something by having two of these have the same score. At a high level, it seems like all we're asserting if that a closer score 3 beats a further score 3. Is there a different spacing that would prerve that `7 - 3 - 1` score layout and get us the desired resuts at zoom 6?

File: src/cpp_util.cpp (hunk 1 above)
Comment (Nmargolis) [reply src/cpp_util.cpp]: Logically it's highly unlikely (but not impossible) that distRatio could ever be > 1 if the distance were 0.8, since the minimum zoom we're allowing is zoom 6. But the C++ compiler doesn't know that and gives this error:
```
../src/cpp_util.cpp:173:5: error: expected expression
    else if (distRatio > 1.0) {
    ^
1 error generated.```

Also, it IS possible, for example if the radius were something like 10 or 20:
distRatio for distance 0.8 zoom 6 radius 10 = `2.56289`
distRatio for distance 0.8 zoom 6 radius 20 = `1.28145`

File: test/coalesce.test.js (hunk 3 above)
Comment (Nmargolis) [reply test/coalesce.test.js]: yup, I was worried about this too so I added additional tests with that score layout below: https://github.com/mapbox/carmen-cache/pull/145/files/3b9d976935af024fa450264ccfaae25fcc73b38e#diff-88079f04b5bc6e5d48f655051dc944a9R431

To be honest, I'm not sure what the "desired results" are, so the expectations for the tests below are that the feature with a score of 7 beats a closer feature with a score of 3. We could adjust the scores a bit to see at what point a closer, lower scored feature beats a further, higher scored feature.

File: test/coalesce.test.js (hunk 2 above)
Comment (Nmargolis) [reply test/coalesce.test.js]: good catch! will make them all consistent to zoom 6

File: test/coalesce.test.js (hunk 3 above)
Comment (Nmargolis) [reply test/coalesce.test.js]: per 🗣, the two fixtures we're comparing are not actually the same. The first and 2nd expected results have different scores, and the 2nd and 3rd have different relevance. Relevance is updated in the fixture below with the old score distribution to compare apples to apples.
